File: code/julia/binary-search.jl
using(Test)

"""
Returns the index position of the target value from the given list
"""
function binary_search(input::Vector, target::Int)::Union{Int, Nothing}
    first = 1
    last = length(input)

    while first <= last
        mp = fld((first + last), 2)

        if input[mp] == target return mp end

        if input[mp] < target
            first = mp + 1
        else
            last = mp - 1
        end

    end

    return nothing

end

input = collect(range(1, 500))
result = binary_search(input, 500)
@test result == 500

input = collect(range(1, 1000))
result = binary_search(input, 1001)
@test isnothing(result)

File: code/py/binary-search.py
def binary_search(list:list, target):
    """
    Returns the index position of the target value from the given list
    """
    first = 0
    last = len(list) - 1
    
    while first <= last:
        midpoint = (first + last) // 2
        
        if list[midpoint] == target:
            return midpoint
        elif list[midpoint] < target:
            first = midpoint + 1
        else:
            last = midpoint - 1
    
    return None

input = range(0, 1001)
result = binary_search(input, 1000)
assert result == 1000

input = range(0, 1001)
result = binary_search(input, 2000)
assert result is None